topics/cluster-tutorial.md

```diff
@@ -928,16 +928,53 @@ it with an updated version of Valkey. If there are clients scaling reads using
 replica nodes, they should be able to reconnect to a different replica if a given
 one is not available.
 
-Upgrading primaries is a bit more complex, and the suggested procedure is:
-
-1. Use `CLUSTER FAILOVER` to trigger a manual failover of the primary to one of its replicas.
-   (See the [Manual failover](#manual-failover) in this topic.)
-2. Wait for the primary to turn into a replica.
-3. Finally upgrade the node as you do for replicas.
-4. If you want the primary to be the node you just upgraded, trigger a new manual failover in order to turn back the upgraded node into a primary.
-
-Following this procedure you should upgrade one node after the other until
-all the nodes are upgraded.
+Upgrading primaries is a bit more complex. The suggested procedure is to trigger
+a manual failover to turn the old primary into a replica and then upgrading it.
+
+A complete rolling upgrade of all nodes in a cluster can be performed by
+repeating the following procedure for each shard (a primary and its replicas):
+
+1. Add one or more upgraded nodes as new replicas to the primary. This step is
+   optional but it ensures that the number of replicas is not compromised during
+   the rolling upgrade. An alternative is to upgrade one replica at a time and
+   have fewer replicas online during the upgrade. To add a new node, use
+   [`CLUSTER MEET`](../commands/cluster-meet.md) and [`CLUSTER
+   REPLICATE`](../commands/cluster-replicate.md) or use `valkey-cli` as
+   described under [Add a new node as a replica](#add-a-new-node-as-a-replica).
```

Comment:
Should we structure this as approach 1 (Existing set of nodes)/approach 2 (Additional node for better availability guarantees)? Or let's add a newline for the alternative suggestion.

Reply:
Do you mean repeat the full 7 steps for the 2 approaches?

I'd prefer to keep them unified so I'll just add a newline. You could actually do any mix of the to approaches (add some, upgrade some, remove some). We replace all the nodes and remove all the old ones. I think it depends whether you use VMs or pods, and whether you use persistence or not.

Reply:
I've added a blank line between the alternatives here.